docs/shared/recipes/generators/create-sync-generator.md

```diff
@@ -0,0 +1,136 @@
+# Create a Sync Generator
+
+Sync generators are generators that are used to ensure that your file system is in the correct state before a task is run or the CI process is started. From a technical perspective, a sync generator is no different from any other generator, but it has some additional performance considerations and needs to be registered in a particular way.
+
+## Create a Generator
+
+Create a sync generator the same way you would [create any generator](/extending-nx/recipes/local-generators):
+
+```shell
+nx g my-sync-generator --directory=tools/my-plugin/src/generators/my-sync-generator
+```
+
+A sync generator should be able to run without any required options, so update the schema accordingly:
+
+```jsonc {% fileName="tools/my-plugin/src/generators/my-sync-generator/schema.json" %}
+{
+  "$schema": "https://json-schema.org/schema",
+  "$id": "MySyncGenerator",
+  "title": "",
+  "type": "object",
+  "properties": {},
+  "required": []
+}
+```
+
+Sync generators can optionally return an `outOfSyncMessage` to display to users when the sync generator needs to be run.
+
+```ts {% fileName="tools/my-plugin/src/generators/my-sync-generator/my-sync-generator.ts" %}
+import { Tree } from '@nx/devkit';
+import { SyncGeneratorResult } from 'nx/src/utils/sync-generators';
+
+export async function mySyncGenerator(
+  tree: Tree
+): Promise<SyncGeneratorResult> {
+  if (
+    !tree.exists('/legal-message.txt') ||
+    tree.read('/legal-message.txt').toString() !==
+      'This is an important legal message.'
+  ) {
+    tree.write('/legal-message.txt', 'This is an important legal message.');
+  }
+  return {
+    outOfSyncMessage: 'The legal-message.txt file needs to be created',
+  };
+}
+
+export default mySyncGenerator;
+```
+
+## Use the Project Graph
+
+One of the benefits of sync generators is that you can keep configuration files up to date with your project graph. To read from the project graph, use the [`createProjectGraphAsync`](/nx-api/devkit/documents/createProjectGraphAsync) from the `@nx/devkit` package. It could look something like this:
+
+```ts {% fileName="tools/my-plugin/src/generators/my-sync-generator/my-sync-generator.ts" %}
+import { Tree, createProjectGraphAsync } from '@nx/devkit';
+import { SyncGeneratorResult } from 'nx/src/utils/sync-generators';
+
+export async function mySyncGenerator(
+  tree: Tree
+): Promise<SyncGeneratorResult> {
+  const projectGraph = await createProjectGraphAsync();
+  const numberOfProjects = Object.values(projectGraph.nodes).length;
+  tree.write(
+    '/project-count.txt',
+    `There are ${numberOfProjects} projects in the repository.`
+  );
+  return {
+    outOfSyncMessage: 'The project-count.txt file needs to be created',
```

Comment:
Nit: I know it's just an example but this would be a more appropriate message given it will also be out of sync when the project count changes.

```suggestion
    outOfSyncMessage: 'The project-count.txt file needs to be update',
```